[WebCore] Shrink FontDescription

tadeuzagallo opened the pull request:
#### 26968a0b0a522e031d2d2e5887d1e16f089caecb
<pre>
[WebCore] Shrink FontDescription
<a href="https://bugs.webkit.org/show_bug.cgi?id=252478">https://bugs.webkit.org/show_bug.cgi?id=252478</a>
rdar://105594136

Reviewed by Simon Fraser.

Use `WTF::Marked` instead of `std::optional` in `FontDescription::sizeAdjust` to
reduce its size from 216 to 208 bytes. Consequently, this shrinks `StyleRareData`
from 320 to 304 bytes, which takes it from the 352b to the 304b size class.

* Source/WebCore/platform/graphics/FontCascadeDescription.cpp:
* Source/WebCore/platform/graphics/FontDescription.h:
(WebCore::FloatMarkableTraits::isEmptyValue):
(WebCore::FloatMarkableTraits::emptyValue):

Canonical link: <a href="https://commits.webkit.org/260564@main">https://commits.webkit.org/260564@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/26997d769f04b96241105cee3c4096ad467dc8c7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108341 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17435 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41205 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117457 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116832 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112227 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18942 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8714 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100563 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114109 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14203 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97380 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42111 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96181 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29024 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83791 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10269 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30368 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11011 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7275 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16419 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49965 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7304 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12597 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/platform/graphics/FontDescription.h:
```diff
@@ -31,12 +31,25 @@
 
 using namespace WebKitFontFamilyNames;
 
+struct FloatMarkableTraits {
```
Should we share this in `Markable.h` ?

`Markable<float>` will be useful in plenty of situation on the codebase.

tadeuzagallo replied:
I considered that, the only thing that me question it is because we can't pass the empty value through a template argument, since it's disallowed to pass floats as arguments, so it makes the traits a little less flexible. Maybe I can still move it there and somehow make it explicit from the name that it'll use NAN as the empty value

tadeuzagallo committed:
[WebCore] Shrink FontDescription
https://bugs.webkit.org/show_bug.cgi?id=252478
rdar://105594136

Reviewed by Simon Fraser.

Use `WTF::Marked` instead of `std::optional` in `FontDescription::sizeAdjust` to
reduce its size from 216 to 208 bytes. Consequently, this shrinks `StyleRareData`
from 320 to 304 bytes, which takes it from the 352b to the 304b size class.

* Source/WebCore/platform/graphics/FontCascadeDescription.cpp:
* Source/WebCore/platform/graphics/FontDescription.h:
(WebCore::FloatMarkableTraits::isEmptyValue):
(WebCore::FloatMarkableTraits::emptyValue):

Canonical link: https://commits.webkit.org/260564@main
